### PR TITLE
 refactor(requestv2): extract request handler methods

### PR DESCRIPTION
**Pull request by drakulis**
regarding #1664. Refactored third point from the list.

- [x] extract common `createResponse` method to avoid adding `XML_HEADER`, `SOAP_ENCODING_HEADER` and `SOAP_ENCODING_FOOTER` over and over again
- [x] extract `searchHandler` method
- [x] extract other easy to extract handlers like `imageHandler`, `mediaReceiverRegistrarHandler`, `getSystemUpdateIdHandler` and others
- [ ] `inputStream` as a local variable instead of class property. (required for further refactor)
- [ ] create separate methods for `streamResponse`, `xmlResponse`, `headerResponse` 
- [ ] extract fileHandler (movie, music etc). This one will be tricky.
- [ ] ...
- [ ] extract service handlers (`ContentDirectoryHandler`, `ConnectionManagerHandler` etc)
- [ ] drop support for first version of request (`Request` class)

### CODE REVIEW

File: src/main/java/net/pms/network/RequestV2.java [hunk 1]
```diff
@@ -618,89 +618,30 @@ public ChannelFuture answer(
...
-				output.headers().set(HttpHeaders.Names.CONTENT_TYPE, "image/jpeg");
-			}
-
-			output.headers().set(HttpHeaders.Names.ACCEPT_RANGES, HttpHeaders.Values.BYTES);
-			output.headers().set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.KEEP_ALIVE);
-			output.headers().set(HttpHeaders.Names.EXPIRES, getFUTUREDATE() + " GMT");
-			inputStream = getResourceInputStream(argument);
+			inputStream = imageHandler(output);
 		} else if ((method.equals("GET") || method.equals("HEAD")) && (argument.equals("description/fetch") || argument.endsWith("1.0.xml"))) {
 			output.headers().set(HttpHeaders.Names.CONTENT_TYPE, "text/xml; charset=\"utf-8\"");
```
Comment (SubJunk) [inline src/main/java/net/pms/network/RequestV2.java]: @drakulis  I'm sure there is a good reason for this but why are these lines being left out of the new methods?

Comment (drakulis) [reply src/main/java/net/pms/network/RequestV2.java]: @SubJunk Yes, as you see we have nice pattern here. Take a look, I think 80% of requests are setting this header this way. It means that `CONTENT_TYPE` should be by default set to `"text/xml; charset=\"utf-8\"` and rest of requests should have own specific `CONTENT_TYPE`. I want to clean it up in some next batch of changes

Comment (SubJunk) [reply src/main/java/net/pms/network/RequestV2.java]: Cool that sounds good

**Comment by SubJunk** [comment]
@drakulis it seems that this broke browsing for me on my Panasonic TV so I will do an urgent hotfix release to revert it as soon as I can, unless you are able to fix it before then

**Comment by SubJunk** [comment]
It's working again in https://github.com/UniversalMediaServer/UniversalMediaServer/pull/1687

**Comment by drakulis** [comment]
I'm using UMS with my changes since few days and it seems to work.

@SubJunk do you see any error message in the logs? 

I will check  VLC when I'm back home.

**Comment by SubJunk** [comment]
@drakulis I'll check the logs after I finish this release if I have time, I'm doing it now

**Comment by SubJunk** [comment]
@drakulis the release is done now, I'll get logs tomorrow if you still need them. Also just so you know I take full responsibility for this, I should have done manual testing on it.
Luckily there was only around 8 hours between releases and roughly 2k downloads so the damage was fairly limited

**Comment by drakulis** [comment]
@SubJunk please provide logs. I think not all users were affected. As I wrote before I'm using UMS with my changes since few day and I don't have browsing issue - checked on TV, on mobile and VLC (on windows 10) and there is not problems either. Also pipin77 during tests of #1676 were using version with extracted `searchHandler` and didn't report such issue

What are the symptoms of this issue on your side? Do you have empty list of files? App is crushing?

**Comment by drakulis** [comment]
@SubJunk I've just made a clean install of v7.6.1. It works without any issue on all my tested devices. Could it be related with some database entries? java? network issues? firewall? I have no idea why it is not working on your side. Looking forward for logs (loglevel=TRACE ;) )

**Comment by drakulis** [comment]
@SubJunk yes, sure, I'm aware that VLC on windows and iOS have different implementations. But at the end it is just a matter of protocol which has to be handled. If it is broken, it should be broken on all clients (or most of them).
I made additional tests on my android tablet (LG) with following apps: "DMS Explorer", "MediaHouse UPnP", "VLC for Android", "Slick UPnP". All are working fine without any issues.

Therefore, I'm suspecting some temporary network problems on your side. But yeah, lets take a look on logs.

**Comment by ExSport** [comment]
Thanks for Slick which offered me also **UPnP Tool** similar to **Intel Developer Tools** or **Cling** (or what was the tool name, now not available, with similar analyser like this one) on PC.
https://play.google.com/store/apps/details?id=com.tjjang.upnptool

**Comment by SubJunk** [comment]
@drakulis The error in the logs seems to be:

```
java.lang.NoClassDefFoundError: javax/xml/soap/MessageFactory
	at net.pms.network.RequestV2.getPayload(RequestV2.java:1273)
	at net.pms.network.RequestV2.browseHandler(RequestV2.java:1053)
	at net.pms.network.RequestV2.answer(RequestV2.java:646)
	at net.pms.network.RequestHandlerV2.writeResponse(RequestHandlerV2.java:382)
	at net.pms.network.RequestHandlerV2.messageReceived(RequestHandlerV2.java:266)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.handler.stream.ChunkedWriteHandler.handleUpstream(ChunkedWriteHandler.java:142)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.handler.codec.http.HttpChunkAggregator.messageReceived(HttpChunkAggregator.java:145)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:485)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

[7.6.1-brokenbrowse.zip](https://github.com/UniversalMediaServer/UniversalMediaServer/files/2609785/7.6.1-brokenbrowse.zip)

**Comment by Nadahar** [comment]
@drakulis I don't know what is causing this particular error, from the log above it can look like some relatively simple Java version / dependency error. That said, I feel for the situation you're in since I've been in this situation many times.

Since I joined UMS in 2015, I've been more or less the only one that has attempted to refactor things to "get the code under control". It's been a difficult and lonely path. What you can expect from the others (in my experience) is nothing/silence and a general lack of interest as long as everything goes well, and a lot of blame if anything goes wrong, even if what you did just revealed some other bug.

On top of that it's the state of the code itself. Since the people who originally wrote much of the code are long gone (PMS era), little or no effort have been put into actually understanding the code when making changes or "bug fixes". This means that most of what has been done to existing classes for many years no are mostly "hacks". By "hacks" I mean, changes that manipulates the existing logic without really looking at the consequences, often using "hardcoded tricks" or custom parameters (so that it's easy to disable when it turns out it broke stuff) with little or no testing. This obviously leads to a lot of bugs, which is then again fixed in the same way. Some classes in the codebase suffer from a lot of layers of such hacks, so much so that when I'm trying to figure out what a piece of code is supposed to do, I often end up having to walk far back in Git history (2010 - 2013) to see what it was originally designed to do.

Since this way of doing things removes all logical structure from the code, every change can break something that is supposedly completely unrelated. Some hacks rely on other hacks or bugs, so that if you the "parent bugs" you will break the "child bugs". It's an ugly situation, and it's what finally made me give up on UMS and start DMS instead.

There are certain classes I think of as "unrefactorable" as it is, and which I believe must first carefully by "offloaded" from as much work as possible before they can be refactored. Network, NetworkRequest, DLNAResource, RendererConfiguration, FFMpegVideo, MEncoderVideo are the ones I can think of at the moment, but I know there are more.

I really like what you are trying to do and hope you'll succeed, but I hope you have a thick skin and a lot of persistence :wink:

**Comment by drakulis** [comment]
@Nadahar Thanks for your support and understanding. I was trying to participate PMS long time ago but project seems to be already dead then. So I also gave up, made some changes for myself and was using modified PMS till now on. Now when UMS seems to be actively developed I think it is worth to try again and give something from myself to community. Legacy PMS code is terribly designed and I'm getting headache when I have to look into it so I'm very determined to clean it up ;) - maybe not everything but at least areas where I'm making changes.

As I see there are just few active developers and lots of environments and devices to support, so yes, shit may happens ;) as it is in this case. We may just get a lesson from it and together figure out how to prevent such situations in the future.

@SubJunk 
It is definitely Java JRE/JDK issue. I'm using Java 7 to build the project and Java 8 to run it on my server. As I see you are running UMS on Java 10. I have googled a bit and found following:
```
JavaSE 9 moved package javax.xml.soap to the module java.xml.ws.
A quick workaround is to either
- run the jar with JRE 8: $MY_JRE8_HOME/bin/java -jar my.jar, or
- add a module for JRE 9: java --add-modules java.xml.ws -jar my.jar
```
So in Java 10 this package is also in module which seems to has be explicitly enabled. Thats really sucks!

But I think we may use external jar to add support for this package: https://mvnrepository.com/artifact/javax.xml.soap/javax.xml.soap-api and be independent from Java version

I will add it to the project and do some test on various versions of java. If it works I will prepare PR again.

**Pull request merged**